File: scripts/analyze_data.py
import pandas as pd
import numpy as np
from functools import reduce
import os
import scipy.stats as stats
from sklearn import linear_model
import matplotlib.pyplot as plt
import seaborn as sns
import hvplot.pandas
from pingouin import pairwise_tukey

# drop Hampden district and save district totals
statetotals = pd.DataFrame(data.iloc[406])
statetotals = statetotals.transpose()
data = data[0:406]

#-------------------------#
#EXPLORATORY DATA ANALYSIS#
#-------------------------#

#shape
data.shape

## MISSING DATA ##

# counting and visualizing missing data
nalist = list()
for name in data:
    total = (data[name] == 99999).sum()
    nalist.append(total)
plt.bar(range(len(data.columns)), nalist)
plt.ylabel('Frequency')
plt.xlabel('Column Index')
plt.title("Frequency of Missing Data per Column")
plt.show()



## ENROLLMENT DATA ##

# counting and visualizing enrollment data with histogram
enrollment = pd.Series(data['Total'], name="total no. students")
ax = sns.distplot(enrollment).set_title('distribution of total enrollment by district')
np.median(enrollment)
np.std(enrollment)
#   boston outlier
np.max(enrollment)
enrollment.skew()



## COLLEGE ATTENDANCE RATES FOR ALL, WHITE, BLACK, HISPANIC ##

# all
collegeattendall = pd.Series(data['Attending Coll./Univ. (%)'], name="all college attendance %")
collegeattendall = pd.Series([i for i in collegeattendall if i != 99999])
np.mean(collegeattendall)
np.std(collegeattendall)
collegeattendall.skew()
collegeattendall = pd.DataFrame(collegeattendall)
collegeattendall['Percent'] = collegeattendall.iloc[:,0]
collegeattendall['Ethnicity'] = 'All'
stat, p = stats.shapiro(collegeattendall)

#black
collegeattendblack = pd.Series(data['Black Attending Coll./Univ. (%)'], name="black college attendance %")
collegeattendblack = pd.Series([i for i in collegeattendblack if i != 99999])
np.mean(collegeattendblack)
np.std(collegeattendblack)
collegeattendblack.skew()
collegeattendblack = pd.DataFrame(collegeattendblack)
collegeattendblack['Percent'] = collegeattendblack.iloc[:,0]
collegeattendblack['Ethnicity'] = 'Black'

#hispanic
collegeattendhisp = pd.Series(data['Hisp Attending Coll./Univ. (%)'], name="hisp college attendance %")
collegeattendhisp = pd.Series([i for i in collegeattendhisp if i != 99999])
np.mean(collegeattendhisp)
np.std(collegeattendhisp)
collegeattendhisp.skew()
collegeattendhisp = pd.DataFrame(collegeattendhisp)
collegeattendhisp['Percent'] = collegeattendhisp.iloc[:,0]
collegeattendhisp['Ethnicity'] = 'Hispanic/Latino'

#white
collegeattendwhite = pd.Series(data['White Attending Coll./Univ. (%)'], name="white college attendance %")
collegeattendwhite = pd.Series([i for i in collegeattendwhite if i != 99999])
np.mean(collegeattendwhite)
np.std(collegeattendwhite)
collegeattendwhite.skew()
collegeattendwhite = pd.DataFrame(collegeattendwhite)
collegeattendwhite['Percent'] = collegeattendwhite.iloc[:,0]
collegeattendwhite['Ethnicity'] = 'White'

# economocially disadvantaged
collegeattendlowecon = pd.Series(data['Low Income Attending Coll./Univ. (%)'], name="low income college attendance %")
collegeattendlowecon = pd.Series([i for i in collegeattendlowecon if i != 99999])
np.mean(collegeattendlowecon)
np.std(collegeattendlowecon)
collegeattendlowecon.skew()
collegeattendlowecon = pd.DataFrame(collegeattendlowecon)
collegeattendlowecon['Percent'] = collegeattendlowecon.iloc[:,0]
collegeattendlowecon['Ethnicity'] = 'Low Income'

# boxplot of all
collegeattend = pd.concat([collegeattendall,collegeattendhisp,collegeattendwhite,collegeattendblack,collegeattendlowecon])
ax = sns.boxplot(data=collegeattend,x='Ethnicity',y='Percent').set_title('Boxplot of College Attendance by Ethnicity')

# one way ANOVA for hispanic vs. other races for college attendance
stats.f_oneway(collegeattendwhite,collegeattendhisp,collegeattendblack)
# creating model
model = ols('Percent ~ C(Ethnicity)', data=collegeattend).fit()
anova_table = sm.stats.anova_lm(model, typ=2)
anova_table
# multiple comparisons
m_comp = pairwise_tukey(data=collegeattend, dv='Percent', between='Ethnicity')
print(m_comp)
# checking ANOVA assumptions
w, pvalue = stats.shapiro(model.resid)
print(w, pvalue)


# one way anova non-parametric
stat, p = stats.kruskal(collegeattendall, collegeattendblack, collegeattendhisp)


## LOOKING AT PREDICTOR VARIABLES ##


## Potential Predictor Variables

# maybe add AP results
predictorvars = data.loc[:, ['enroll_low_econ',"Average Class Size","Total Expenditures per Pupil", "Average Salary", '% All Completed Advanced', 'reading_writing_all', 'math_all', '% Exemplary','% Proficient', '% Needs Improvement', 'Teacher % Retained','Low Income Attending Coll./Univ. (%)']]
predictorvarshisp = data.loc[:, ['%_staff_hispanic','enroll_hispanic','Percent Class Hispanic',"Average Class Size","Total Expenditures per Pupil", "Average Salary", '% All Completed Advanced', 'reading_writing_all', 'math_all', '% Exemplary','% Proficient', '% Needs Improvement', 'Teacher % Retained','Hisp Attending Coll./Univ. (%)']]
hey2 =  predictorvars[(predictorvars.iloc[:, 1:] != 99999).all(axis=1)]
heycorr = hey2.corr()
heycorr


# visualizing funding per student with histogram

fundingperstudent = pd.Series(data['Total Expenditures per Pupil'], name="total expenditures per pupil")
fundingperstudent = pd.Series([i for i in fundingperstudent if i != 99999])
ax = sns.distplot(fundingperstudent).set_title('distribution of per pupil expenditure by district')
np.mean(fundingperstudent)
np.std(fundingperstudent)
np.min(fundingperstudent)
np.max(fundingperstudent)
fundingperstudent.skew()

# average salaries

avgsal = pd.Series(data['Average Salary'], name="Average Salary")
avgsal = pd.Series([i for i in avgsal if i != 99999])
ax = sns.distplot(avgsal).set_title('average salary by district')
np.mean(avgsal)
np.std(avgsal)
np.min(avgsal)
np.max(avgsal)
avgsal.skew()

# sat score distributions (all)

# sat score distributions (non-white)

# sat score distributions (female)


#-------------------------#
#EXPLORATORY DATA ANALYSIS#
#-------------------------#

## FIND CORRELATIONS BETWEEN ALL PAIRS OF INTERESTING COLUMNS, BUT MUST SUBSET DATA FIRST ##


# advancednonwhite(money)
nonwhiteadvanced = data.loc[:, ["%_completedadvanced_nonwhite","Total Expenditures per Pupil","Average Salary"]]
indexNames = nonwhiteadvanced[nonwhiteadvanced["%_completedadvanced_nonwhite"] == 99999].index
nonwhiteadvanced.drop(indexNames , inplace=True)
nonwhiteadvanced.hvplot(x="%_completedadvanced_nonwhite", y=["Total Expenditures per Pupil","Average Salary"], kind='scatter',ylabel='USD',legend='right',xlabel='% non-white students completing advanced courses')

nonwhiteadvancedcorr = nonwhiteadvanced.corr()
nonwhiteadvancedcorr

File: scripts/analyze_data.R
library(tidyverse)
library(ggpubr)
library(rstatix)
library(car)
library(moments)
library(readr)
library(Hmisc)
library(corrplot)

# importing data 
data <- read_csv("/Users/zacharycollester/Documents/ma_public_schools/data/csv_R/data.csv")

# getting state total data
statetotals <- data.frame(data[407,])
# removing state total and Hampden school from dataframe
data <- data[-c(407,408),]

## ENROLLMENT DATA ## 
# missing data
nalist <- vector()
for (i in 1:ncol(data)){
  total <- sum(data[,i] == 99999)
  nalist[i] <- total
}
nalist <- data.frame(nalist)
missingdata <- ggplot(nalist, aes(x=nalist)) + geom_histogram(color="black", fill="lightblue") + 
  geom_vline(aes(xintercept=mean(nalist)),color="darkblue", linetype="dashed", size=1) +
  labs(y="Frequency",x="Number of Missing Data Points") +
  ggtitle("Histogram of Missing Data")
missingdata

# total enrollment per district
enrollment <- vector()
for (i in 1:nrow(data)){
  enrollment[i] <- as.integer(data[i,"Total"])
}
enrollment <- data.frame(enrollment)
enrollmentdata <- ggplot(enrollment, aes(x=enrollment)) + geom_histogram(color="black", fill="lightblue") + 
  geom_vline(aes(xintercept=median(enrollment)),color="darkblue", linetype="dashed", size=1) +
  labs(y="Frequency",x="Number of Students") +
  ggtitle("Histogram of Total Enrollment by District")
enrollmentdata

## COLLEGE ATTENDANCE ##
# all
college <- data.frame(data$`Attending Coll./Univ. (%)`,'All')
# white
collegewhite <- data.frame(data$`White Attending Coll./Univ. (%)`,'White')
# black
collegeblack <- data.frame(data$`Black Attending Coll./Univ. (%)`,'Black')
# hisp
collegehisp <- data.frame(data$`Hisp Attending Coll./Univ. (%)`,'Hispanic/Latino')
# low income
collegeecon <- data.frame(data$`Low Income Attending Coll./Univ. (%)`,'Low Income')
# changing column names
dflist = list(college=college, collegewhite=collegewhite, collegeblack=collegeblack, collegehisp=collegehisp, collegeecon=collegeecon)
colnames <- c("Percent", "Group")
list2env(lapply(dflist, setNames, colnames), .GlobalEnv)
# rbinding all groups, removing null data
collegeattend <- rbind(college,collegewhite,collegeblack,collegehisp,collegeecon)
collegeattend <- filter(collegeattend, Percent != 99999)
# boxplot
collegeboxplot <- ggplot(collegeattend, aes(x=Group, y=Percent, fill=Group)) + 
  geom_boxplot(alpha=0.3) + theme(legend.position="none") +
  scale_fill_brewer(palette="BuPu") +
  ggtitle("Boxplot of College Attendance \nby Racial/Socioeconomic Group")
collegeboxplot

# Scatter plot of enrollment vs college attendance
enrollattend <- data.frame(data$`District Name`,data$Total,data$`Attending Coll./Univ. (%)`, data$`White Attending Coll./Univ. (%)`,
                           data$`Black Attending Coll./Univ. (%)`, data$`Hisp Attending Coll./Univ. (%)`, 
                           data$`Low Income Attending Coll./Univ. (%)`)
colnames <- c('distname','enroll','all','white','black','hisp','lowincome')
names(enrollattend) <- colnames
enrollattend <- enrollattend %>% filter(all != 99999)
ggplot(enrollattend, aes(x=enroll, y=all)) + geom_point() + geom_hline(aes(yintercept=median(all)),color="darkblue", linetype="dashed", size=1)
enrollattend <- enrollattend %>% filter(hisp != 99999)
ggplot(enrollattend, aes(x=enroll, y=hisp)) + geom_point() + geom_hline(aes(yintercept=median(hisp)),color="darkblue", linetype="dashed", size=1)
enrollattend <- enrollattend %>% filter(lowincome != 99999)
ggplot(enrollattend, aes(x=enroll, y=lowincome)) + geom_point() + geom_hline(aes(yintercept=median(lowincome)),color="darkblue", linetype="dashed", size=1)
enrollattend <- enrollattend %>% filter(black != 99999)
ggplot(enrollattend, aes(x=enroll, y=black)) + geom_point() + geom_hline(aes(yintercept=median(black)),color="darkblue", linetype="dashed", size=1)
enrollattend <- enrollattend %>% filter(white != 99999)
ggplot(enrollattend, aes(x=enroll, y=white)) + geom_point() + geom_hline(aes(yintercept=median(white)),color="darkblue", linetype="dashed", size=1)


# Summary Statistics for College Attendance
collegeattend %>% 
  group_by(Group) %>% 
  summarise(count = n(),
            min = min(Percent),
            max = max(Percent),
            mean = mean(Percent),
            median = median(Percent),
            std = sd(Percent))

# State Total Summary Statistics 
statetotals %>% 
  summarise(all = Attending.Coll..Univ......1,
            white = White.Attending.Coll..Univ......1,
            black = Black.Attending.Coll..Univ......1,
            hispanic = Hisp.Attending.Coll..Univ......1,
            lowincome = Low.Income.Attending.Coll..Univ......1)

# Enrollment Outlier Summary Statistics
enrollment_outliers <- data %>% filter(Total > 8000)
# getting rid of lawrence
enrollment_outliers <- enrollment_outliers[-5,]
enrollment_outliers %>% 
  summarise(all = mean(`Attending Coll./Univ. (%)`),
            white = mean(`White Attending Coll./Univ. (%)`),
            black = mean(`Black Attending Coll./Univ. (%)`),
            hispanic = mean(`Hisp Attending Coll./Univ. (%)`),
            lowincome = mean(`Low Income Attending Coll./Univ. (%)`)) 

# Shapiro Test of Normality for College Attendance (not normal)      
collegeattend %>%
  group_by(Group) %>%
  shapiro_test(Percent)

# Levene Test for Homogeneity of Variances (variance is homogenous)
leveneTest(Percent ~ Group, data = collegeattend)

# Kruskall Wallace Test
kruskal.test(Percent ~ Group, data = collegeattend)
# Multiple Comparisons with Wilcox Test
pairwise.wilcox.test(collegeattend$Percent, collegeattend$Group,
                     p.adjust.method = "BH")


# Dataframe of potential predictors (all)
collegeattendvars <- data.frame(data$`Average Class Size`, data$`Average Salary`, data$`Total Expenditures per Pupil`,
                                data$reading_writing_all, data$math_all, data$`% All Completed Advanced`, data$`% Exemplary`,
                                data$`% Proficient`, data$`% Needs Improvement`, data$`Attending Coll./Univ. (%)`)
collegeattendvars <- collegeattendvars %>% filter_all(all_vars(.!= 99999))
columns <- c('class_size', 'avg_salary', 'per_pupil$', 'sat_rr', 'sat_math', 'advanced_course', 'exemplary', 'proficient', 'needs_improv', 'college')
names(collegeattendvars) <- columns
newdata <- collegeattendvars[,-1]

# Correlation Matrix of Predictor Variables (with pearson R2 values)
res <- rcorr(as.matrix(collegeattendvars), type = "pearson")
col <- colorRampPalette(c("#BB4444", "#EE9988", "#FFFFFF", "#77AADD", "#4477AA"))
corrplot(res$r, method = "color", col = col(200),
         type = "upper", order = "hclust", number.cex = .7,
         addCoef.col = "black", # Add coefficient of correlation
         tl.col = "black", tl.srt = 45, # Text label color and rotation
         # Combine with significance
         p.mat = res$P, sig.level = 0.01, insig = "blank", 
         # hide correlation coefficient on the principal diagonal
         diag = FALSE)






# see how correlation matrix changes for larger districts and smaller districts
# multiple regression for big districts (all, hispanic, low income)
# multiple regression for smaller districts (all)
# maybe redo stats and hypothesis testing for small and big district subsets



